packages/astro/src/core/routing/manifest/create.ts

```diff
@@ -476,35 +448,132 @@ export function createRouteManifest(
+	if (a.segments.length !== b.segments.length) {
+		return false;
+	}
+
+	if (a.type === 'fallback' || b.type === 'fallback') {
+		// If either route is a fallback route, they don't collide.
+		// Fallbacks are always added below other routes exactly to avoid collisions.
+		return false;
+	}
+
+	const length = a.segments.length;
+	let hasDynamic = false;
+
+	for (let i = 0; i < length; i++) {
+		const segmentA = a.segments[i];
+		const segmentB = b.segments[i];
+
+		if (segmentA.length !== segmentB.length) {
+			return false;
+		}
+
+		const segmentLength = segmentA.length;
+
+		for (let j = 0; j < segmentLength; j++) {
+			const partA = segmentA[j];
+			const partB = segmentB[j];
+
+			if (
+				// If only one side has a rest parameter, they don't collide
+				partA.spread !== partB.spread
+				// If only one side is dynamic, they don't collide
+				|| partA.dynamic !== partB.dynamic
+			) {
+				return false;
 			}
 
-			// If the existing route is dynamic, prefer the static redirect.
-			const base = path.posix.dirname(existingRoute.route);
-			if (base === redirBase && !dynamicRedir && lastSegmentIsDynamic(existingRoute)) {
-				routes.splice(i, 0, routeData);
-				return;
+			hasDynamic = hasDynamic || partA.dynamic;
+
+			// If both sides are static and have different content, they don't collide
+			if (!partA.dynamic && partA.content !== partB.content) {
+				return false;
 			}
-			i++;
 		}
+	}
 
-		// Didn't find a good place, insert last
-		routes.push(routeData);
+	if (hasDynamic && (a.prerender || b.prerender)) {
+		// If the route is dynamic and either of the routes is prerendered,
+		// we cannot afirm that they collide at this point.
+		// They will only collide if both routes are prerendered and return
+		// the same parameters on `getStaticPaths`, which is checked later.
+		return false;
+	}
+
+	// If every segment and part is equivalent, they collide
+	return true;
+}
+
+/** Create manifest of all static routes */
+export function createRouteManifest(
+	params: CreateRouteManifestParams,
+	logger: Logger,
+): ManifestData {
+	// Create a map of all routes so redirects can refer to any route
+	const routeMap = new Map();
+
+	const projectRoutes = createProjectRoutes(params, logger);
+	projectRoutes.forEach((route) => {
+		routeMap.set(route.route, route);
+	});
+
+	const injectedRoutes = createInjectedRoutes(params);
+	Object.entries(injectedRoutes).forEach(([, routes]) => {
+		routes.forEach((route) => {
+			routeMap.set(route.route, route);
+		});
 	});
+
+	const redirectRoutes = createRedirectRoutes(params, routeMap, logger);
+
+	const routes: RouteData[] = [
+		...[
+			...injectedRoutes['above-project'],
+			...redirectRoutes['above-project'],
+		].sort(comparator),
+		...[
+			...projectRoutes,
+			...injectedRoutes['same-as-project'],
+			...redirectRoutes['same-as-project'],
+		].sort(comparator),
+		...[
+			...injectedRoutes['below-project'],
+			...redirectRoutes['below-project'],
+		].sort(comparator),
+	];
```

Comment:
I don't think using spread operators is the correct choice here. It seems we always create new arrays, which is a hit memory-wise. 

I believe we should mutate the ones we have and create only one.

E.g.
```js
const routes = [];
routes.concat(
	injectedRoutes['above-project'].concat(
		redirectRoutes['above-project']
	).sort(comparator)
);
```

I think this is safe because we aren't going to use `injectedRoutes` and `redirecteRoutes` anymore.

Reply:
I don't think it will make much of a difference. Each `.concat` is a possible allocation while multiple spread operations inside the same `[]` are always a single allocation with the correct size (the sum of the length of all the arrays being spread).

Safety-wise you are correct, using concat is not a safety problem.

I'll try to run some benchmarks for that

Reply:
I stand corrected. `concat` takes half the time of the spread operator, unless the function is called enough times for all JIT passes, then it is just 2% faster (but still consistently faster).

Although I just tested this for 3.5k routes, 500 each of (injected above, redirect before, project, inject with project, redirect with project, injected below, redirect below) and it took about 0.5ms even with JIT disabled, so it might not be a worth optimization at the cost of readability.

I think the code with the spreads is more clear about what would be the final ordering than this:
```ts
const routes = injectedRoutes['above-project']
  .concat(redirectRoutes['above-project'])
  .sort(comparator)
  .concat(
    projectRoutes
      .concat(injectedRoutes['same-as-project'], redirectRoutes['same-as-project'])
      .sort(),
    injectedRoutes['below-project']
      .concat(redirectRoutes['below-project'])
      .sort()
  );
```